---
title: "2_process_gwas_outliers.Rmd"
author: "Andrea Estandia"
date: "16/03/2023"
output: html_document
editor_options: 
  chunk_output_type: console
---
  
```{r setup, echo=FALSE}
# Knitr settings: 
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
options(scipen = 999)
```

```{r, echo=FALSE}
source("./src/0.0_island_rule_source.R")
```

Specify trait
```{r}
trait="body_pc1"
```

Read beagle file, remove allele columns, rename columns and remove extra 
numbers after sample_name
```{r}
outlier <- 
  read.table(file.path(reports_path, 
                       "GWAS", 
                        "outliers",
                       "beagle",
                        paste0(trait,"_outliers.beagle")),
             sep="\t", header=T) %>% 
  select(-c("allele1", "allele2")) %>% 
  as_tibble() %>% 
  pivot_longer(starts_with("Ind")) %>% 
  rename(sample_name=name) %>%
  rename(genotype=value) %>% 
  mutate(sample_name=gsub("\\..*", "", sample_name))
```

Read phenotype file with body size (PC1), pop info and genetic sex
```{r}
pheno <-
  read_tsv(file.path(data_path,
                     "GWAS",
                     "phenotypes",
                      paste0(trait, ".tsv")),
           col_names = FALSE) %>%
  rename(sample_name = X1) %>%
  rename(body_PC1 = X2)

geneticsex <-
  read_csv(file.path(data_path, 
                     paste0("sample_info",
                            "geneticsex.csv"))) %>%
  rename(id = Blood_Number)

pop_info <- 
  read_csv(file.path(data_path,
                     "sample_info",
                     "curated_dataset.csv")) %>%
  rename(x = sample_name) %>%
  rename(sample_name = id) %>%
  rename(id=x) %>% 
  select(sample_name, subspecies,id, pop, region, latitude, longitude) %>% 
  left_join(geneticsex)
```

Generate a vector repeating the genotypes AA, AB, and BB
Paste vector to main beagle dataframe
```{r}
geno <-
  rep(c("AA", "AB", "BB"),
      times = length(outlier$sample_name) / 3)

df <- 
  cbind(outlier, geno)

df <- df %>% 
  filter(genotype>0.34) %>% 
  group_by(sample_name, marker) %>% 
  top_n(1, genotype)
```

Merge all datasets and separate column marker into chr and position
```{r}
df2 <- 
  df %>% 
  left_join(pop_info, by = "sample_name") %>% 
  left_join(pheno, by = "sample_name") %>% 
  separate(marker,into = c("chr", "position")) %>% 
  mutate(position=as.numeric(position)) %>% 
  unite("marker", chr:position, remove=FALSE)
```

Load association output, apply a threshold and select the top 1 and top12
```{r}
LRTs <- 
  fread(file.path(reports_path,
                "GWAS",
                "lrt",
                paste0(trait, "_gwas_wholegenome_eigen2.lrt0.gz")),
                header = T,
                sep = "\t",
                stringsAsFactors = F) %>%
        mutate(Chromosome = str_replace(Chromosome, "chr", "")) %>%
        mutate(h=beta+1.96*SE) %>%
        mutate(u=beta-1.96*SE)
      
order_chr <- as.character(seq(from = 5, to = 30))
reference = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z")
ref = factor(as.factor(reference), levels = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z"))
LRTs <- LRTs[order(factor(as.factor(LRTs$Chromosome), levels = ref)),]
LRTs$n <- 1:nrow(LRTs)

bonf=-log10(0.0001 / nrow(LRTs))

info_outliers <-
  LRTs %>% 
  #filter those outliers above bonferroni cut-off
  filter(-log10(P)> bonf)

top_1_outliers <- 
  info_outliers %>% 
  arrange(P) %>% 
  slice(1)

top_12_outliers <- 
  info_outliers %>% 
  arrange(P) %>% 
  slice(1:12)
```

Plot top outlier - Fig S3 C and D
```{r}
top1_pos <- top_1_outliers$Position
top1_major <- paste0(top_1_outliers$Major,top_1_outliers$Major)
top1_majorminor <- paste0(top_1_outliers$Major,top_1_outliers$Minor)
top1_minor <- paste0(top_1_outliers$Minor,top_1_outliers$Minor)

chr <- df2 %>%
  filter(position == top1_pos) %>% 
  ungroup %>% 
  distinct(chr)

body_outlier_top1 <- 
  df2 %>%
  filter(position == top1_pos) %>% 
  drop_na(body_PC1) %>%
  filter(genotype>0.34) %>% 
  ggplot(aes(y = body_PC1, x = geno, col = geno)) +
  geom_violin(trim = FALSE) +
  stat_summary(
    aes(group = geno),
    geom = "pointrange",
    fun.data = mean_cl_normal,
    fun.args=list(conf.int=0.89),
    size = 1)+
  geom_jitter(width = 0.05,
              height = 0.05,
              alpha = 0.8,
              size=2) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(hjust=1, size=text_size, family = "ubuntu"),
    axis.text.y = element_text(size = text_size, family="ubuntu"),
    axis.title = element_text(size = text_size, family="ubuntu"),
    legend.title = element_text(family="ubuntu"),
        panel.grid.major = element_blank()) +
  scale_x_discrete(labels=c("AA" = top1_major, "AB" = top1_majorminor,
                              "BB" = top1_majorminor))+
  scale_color_manual(values = c("#264653", "#2a9d8f", "#e9c46a"),
                     labels=c("AA" = top1_major, "AB" = top1_majorminor,
                              "BB" = top1_majorminor)) +
  labs(title = paste0(chr[1]$chr, ": ", top1_pos),
       y = "Bill size (PC1)\n", x = "\nGenotype", family="ubuntu")
```

Generate figures S4 and S6
```{r}
list_plots <- list()

for (i in unique(info_outliers$Position)){
  tmp <- df2 %>%
      filter(position == i)
  tmp1 <- top_12_outliers %>% 
    filter(Position == i)
  top1_major <- paste0(tmp1$Major,tmp1$Major)
  top1_majorminor <- paste0(tmp1$Major,tmp1$Minor)
  top1_minor <- paste0(tmp1$Minor,tmp1$Minor)
  
  list_plots[[as.character(i)]] <- 
  df2 %>%
  filter(position == i) %>% 
  drop_na(body_PC1) %>%
  filter(genotype>0.34) %>% 
  ggplot(aes(y = body_PC1, x = geno, col = geno)) +
  geom_violin(trim = FALSE) +
  stat_summary(aes(y = body_PC1),
               fun.data = data_summary,
               geom = "pointrange",
               shape = 1) +
  geom_jitter(width = 0.05,
              height = 0.05,
              alpha = 0.8,
              size=2) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(hjust=1, size=text_size, family = "ubuntu"),
    axis.text.y = element_text(size = text_size, family="ubuntu"),
    axis.title = element_text(size = text_size, family="ubuntu"),
    legend.title = element_text(family="ubuntu"),
        panel.grid.major = element_blank()) +
  scale_x_discrete(labels=c("AA" = top1_major, "AB" = top1_majorminor,
                              "BB" = top1_majorminor))+
  scale_color_manual(values = c("#264653", "#2a9d8f", "#e9c46a"),
                     labels=c("AA" = top1_major, "AB" = top1_majorminor,
                              "BB" = top1_majorminor)) +
  labs(title = paste0(unique(tmp[,"chr"])," ", i),
       y = "Bill size (PC1)\n", x = "\nGenotype", family="ubuntu")
  
}

plots <- wrap_plots(list_plots)

ggsave(
  "top12_outliers_effect_bill_PC1.pdf",
  plots,
  path = figures_path,
  device = "pdf",
  width = 15,
  height = 10,
  dpi = 400
)
```

Classify as de novo or SGV
```{r}
# Function to classify mutations
classify_mutations <- function(df) {
  # Check if AB or BB is present in any other population
  is_sgv <- any(df$pop != "Ambae" & df$pop != "Ambrym" & df$pop != "Efate" & df$pop != "Gaua" & df$pop != "Grand_Terre" & df$pop != "Lifou" & df$pop != "Malekula" & df$pop != "Mare" & df$pop != "Ouvea" & df$pop != "Pentecost" & df$pop != "Santo" & df$pop != "Tanna" & df$pop != "Vanua_Lava" & (df$geno == "AB" | df$geno == "BB"))
  
  # Classify mutations
  if (is_sgv) {
    df$classification <- "SGV"
  } else {
    df$classification <- "De novo"
  }
  
  return(df)
}

# Classify mutations in each dataframe
classified_dataframes <- lapply(list_summary_pop, classify_mutations)

# Define the column name you want to extract
column_name <- "classification"

# Create an empty vector to store the first values
first_values <- c()

# Loop through each data frame in the list
for (df_name in names(classified_dataframes)) {
  # Extract the desired column from the current data frame
  extracted_column <- classified_dataframes[[df_name]][[column_name]]
  
  # Get the first value from the extracted column
  first_value <- extracted_column[1]
  
  # Append the first value to the vector
  first_values <- c(first_values, first_value)
}

# Print the merged vector of first values
print(first_values)

table(first_values) 

# Example: Printing the classified dataframes
for (i in seq_along(classified_dataframes)) {
  cat("Dataframe", i, ":\n")
  print(classified_dataframes[[i]])
  cat("\n")
}

for (sublist in 1:length(classified_dataframes)){
  write.xlsx(classified_dataframes[sublist], 
             file=file.path(reports_path,
                            "GWAS",
                            "outliers",
                            "sgv_denovo",
                            "body_PC1_denovo_sgv.xlsx"), 
             sheetName=paste(sublist), 
             row.names=FALSE,
             append=T)
}
```

Prepare file to automatically search in annotation
```{r}
# Function to merge nearby positions within a chromosome
merge_positions <- function(df, threshold) {
  merged_df <- data.frame()
  current_group <- NULL
  
  for (i in 1:nrow(df)) {
    pos <- df$Position[i]
    
    if (is.null(current_group)) {
      current_group <- df[i, ]
    } else {
      if (pos - current_group$Position[nrow(current_group)] <= threshold) {
        current_group <- rbind(current_group, df[i, ])
      } else {
        # Merge current group
        merged_row <- data.frame(Position = c(min(current_group$Position), max(current_group$Position)))
        merged_row <- cbind(merged_row, current_group[1, -(1:2)])
        
        # Append to merged dataframe
        merged_df <- rbind(merged_df, merged_row)
        
        current_group <- df[i, ]
      }
    }
  }
  
  # Add the last group
  if (!is.null(current_group)) {
    merged_row <- data.frame(Position = c(min(current_group$Position), max(current_group$Position)))
    merged_row <- cbind(merged_row, current_group[1, -(1:2)])
    
    merged_df <- rbind(merged_df, merged_row)
  }
  
  merged_df
}

# Specify the threshold for merging (10000 bp)
threshold <- 10000

# Group dataframe by Chromosome and apply merging within each group
merged_df <- info_outliers %>%
  group_by(Chromosome) %>%
  do(merge_positions(., threshold)) %>% 
  distinct(Position,.keep_all=TRUE)

# Remove grouping information
merged_df <- ungroup(merged_df) 

# View the merged dataframe
merged_df 
  
# Create a new dataframe with start_position, end_position, and chromosome_name
df_total <- merged_df %>%
  mutate(start_position = Position - 10000,
         end_position = Position + 10000,
         chromosome = paste0("chromosome",Chromosome)) %>%
  select(start_position, end_position, chromosome)

write.table(df_total, 
            file.path(reports_path,
                      "GWAS",
                      "outliers",
                      "annotation",
                      paste0("positions4annotation_",trait,".csv")), row.names = F,
            sep="\t", quote = FALSE)
```