---
title: "0.0_fig3"
author: "Andrea Estandia"
date: "31/05/2023"
output: html_document
editor_options: 
  chunk_output_type: console
params:
  beagle: 
  - chr27_959841-chr27_1052666.beagle
  - chr28_4071701-chr28_4275666.beagle
  - chr29_1573309-chr29_1664147.beagle
 # - chr30_1054208-chr30_1110074.beagle
---


```{r setup, echo=FALSE}
# Knitr settings: 
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
options(scipen = 999)
```

```{r, echo=FALSE}
source("./src/0.0_island_rule_source.R")
```

Read phenotype file with body size (PC1), pop info and genetic sex
```{r}
pheno <-
  read_tsv(file.path(data_path,
                     "wgs",
                     "lists",
                     "body_pc1.tsv"),
           col_names = FALSE) %>%
  rename(sample_name = X1) %>%
  rename(body_PC1 = X2)

pop_info <- 
  read_csv(file.path(data_path,
                     "phenotypes",
                     "phenotypes+SMC2+AE.csv")) %>%
  rename(x = sample_name) %>%
  rename(sample_name = id) %>%
  select(sample_name, pop, region, latitude, longitude)

geneticsex <-
  read_csv(file.path(data_path, 
                     paste0("phenotypes/geneticsex.csv"))) %>%
  rename(id = Blood_Number)
```

Read results local PCA
```{r}
mds_scores <-
  read.csv(file.path(reports_path,
                       "localPCA/",
                       "unpruned/",
                       "big_data/",
                       "bigdata.csv"),
             header = T)
```

Specify parameters for Fig A and B
```{r}
max_MDS <- 2#nb of mds on whcih to work
sd_lim <- 4#limit for being outlier
x_between <-
  10 #merge window with this number of windows between them
min_n_window <- 5#remove cluster with less than this number of window
unique_lg <- unique(mds_scores$LG)
```

```{r}
MDS_CLUSTER_ALL <- matrix(ncol = 10)
colnames(MDS_CLUSTER_ALL) <-
  c(
    "LG",
    "start",
    "stop",
    "x",
    "window_start",
    "window_stop",
    "n_windows",
    "mds",
    "mds_num",
    "cluster"
  )

order_chr <- as.character(seq(from = 5, to = 30))
reference = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z")
ref = factor(as.factor(reference), levels = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z"))
mds_scores <- mds_scores[order(factor(as.factor(mds_scores$LG), levels = ref)),]
mds_scores$n <- 1:nrow(mds_scores)

axisdf <-
  mds_scores %>%
  group_by(LG) %>%
  summarize(center = (max(n) + min(n)) / 2)

plot_mds <- function(mds_vector, color_values) {
  mds_mean <- mean(mds_vector)
  mds_sd <- sd(mds_vector)
  
  plot <- mds_scores %>%
    ggplot(aes(x = n, y = mds_vector, color = LG)) +
    geom_point() +
    scale_x_continuous(
      label = axisdf$LG,
      breaks = axisdf$center,
      expand = c(0.01, 0.01)
    ) +
    scale_color_manual(values = color_values) +
    geom_hline(yintercept = mds_mean + sd_lim * mds_sd, linetype = "dashed") +
    geom_hline(yintercept = mds_mean - sd_lim * mds_sd, linetype = "dashed") +
    theme_classic() +
    theme(
      text = element_text(size = 12),
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      panel.border = element_blank()
    )
  
  return(plot)
}

mds_vector1 <- mds_scores[, "PC1"]
color_values1 <- rep(c("#8fc7c0", "#2a9d8f"), length(unique_lg))
plot_mds1 <- plot_mds(mds_vector1, color_values1)

mds_vector2 <- mds_scores[, "X2"]
color_values2 <- rep(c("#eddd9a", "#d19f28"), length(unique_lg))
plot_mds2 <- plot_mds(mds_vector2, color_values2)
```

How long are inversions?
```{r}
mds_scores %>% filter(X2>0.4445288) %>% filter(LG=="chr27") %>% dim()
#39 windows*100 SNPs=3900
mds_scores %>% filter(X2>0.4445288) %>% filter(LG=="chr28") %>% dim()
#41 windows*100SNPs=4100
 mds_scores %>% filter(X2>0.4445288) %>% filter(LG=="chr29") %>% dim()
#24 windows*100SNPs=2400
```


Function to create individual MDS plots
```{r}
j=2
plot_mds <- list()
mds_col <- 6 + j
mds_vector <- mds_scores[, mds_col]
mds_mean <- mean(mds_vector)
mds_sd <- sd(mds_vector)

create_mds_plot <- function(dataset, chr) {
  mds_scores_chr <- dataset %>% 
    filter(LG == chr)
  
  if (chr == "chr4A") {
    mds_col <- 5 + j
  } else {
    mds_col <- 6 + j
  }
  mds_vector_chr <- mds_scores_chr[, mds_col]
  
  # calculate limit
  # plot
  if (chr == "chr4A") {
    color_value <- "#2a9d8f"  # Set color to "#2a9d8f" for input "chr4A"
  } else {
    color_value <- "#e9c46a"  # Set the initial color for other inputs
  }
  
  plot_mds_chr <- mds_scores_chr %>%
    ggplot(aes(x = n, y = mds_vector_chr, color = LG)) +
    geom_point() +
    scale_x_continuous(
      label = axisdf$LG,
      breaks = axisdf$center,
      expand = c(0.01, 0.01)
    ) +
    scale_color_manual(values = color_value) +
    geom_hline(yintercept = mds_mean + sd_lim * mds_sd,
               linetype = "dashed") +
    geom_hline(yintercept = mds_mean - sd_lim * mds_sd,
               linetype = "dashed") +
    theme_classic() +
    theme(
      text = element_text(size = 12),
      legend.position = "none",
      panel.border = element_blank(),
      axis.ticks.x = element_blank(),
      axis.text.x.bottom = element_blank()) +
    labs(
      x = paste0("\n", as.character(chr), "\n"),
      y = "MDS2"
    )
  return(plot_mds_chr)
}

```

Plot individual chromosomes
```{r}
list_chr <- c("chr4A", "chr27","chr28","chr29", "chr30")
plot_mds_complete <- list()
for (chr in list_chr) {
  plot_mds_complete[[chr]] <- create_mds_plot(mds_scores, as.character(chr))
}
  
outlier_mds_plots <- wrap_plots(plot_mds_complete)
```

##############################################################################

```{r}
for (par in params$beagle) {
  beagle_file <- file.path(
  reports_path,
    "localPCA",
    "unpruned",
    "outliers",
    "beagle_by_window",
    "max2_merge10_filter5",
    par)

if (!file.exists(beagle_file)) {
gunzip(paste0(beagle_file, ".gz"), remove=FALSE)
}
  
  outlier <- 
  read.table(
    file.path(
      reports_path,
      "localPCA",
      "unpruned",
      "outliers",
      "beagle_by_window",
      "max2_merge10_filter5",
      par
    ),
    header = T
  ) %>%
  select(-c("allele1", "allele2")) %>%
  as_tibble() %>%
  pivot_longer(starts_with("Ind")) %>%
  rename(sample_name = name) %>%
  rename(genotype = value) %>%
  mutate(sample_name = str_remove(sample_name, "\\.[^.]*$"))
  
  geno <-
  rep(c("AA", "AB", "BB"),
      times = length(outlier$sample_name) / 3)

df <- 
  cbind(outlier, geno)

df <- df %>% 
  filter(genotype>0.34) %>% 
  group_by(sample_name, marker) %>% 
  top_n(1, genotype)

df2 <- 
  df %>% 
  left_join(pop_info, by = "sample_name") %>% 
  left_join(pheno, by = "sample_name") %>% 
  separate(marker,into = c("chr", "position")) %>% 
  mutate(position=as.numeric(position)) %>% 
  unite("marker", chr:position, remove=FALSE)

label <- 
  read.table(file.path(data_path,"localPCA/pop_label"))

cov_mat <-
  as.matrix(read.table(file.path(
    reports_path,
      "localPCA",
      "unpruned",
      "outliers",
      "cov_by_window",
      "max2_merge10_filter5",
      paste0(gsub("\\..*","",par), ".cov")
  )))

mds.cor <- (1 - cov_mat) %>%
  cmdscale(k=3, eig = TRUE)

colnames(mds.cor$points) <- c("Dim.1", "Dim.2", "Dim.3")
rownames(mds.cor$points) <-
  label$V3

kmeans_res<-
  kmeans(as.matrix(mds.cor$points[,1]), 
                   c(min(mds.cor$points[,1]), 
                     median(mds.cor$points[,1]), 
                     max(mds.cor$points[,1])))
k_ss<-
  round(kmeans_res$betweenss/kmeans_res$totss,3)

k <- as.data.frame(kmeans_res$cluster)
colnames(k) <- "k"

clusters <- 
  label %>%
  select(-V3, -V4) %>% 
  cbind(mds.cor$points) %>% 
  cbind(k) %>% 
  rename(id=V1) %>% 
  rename(sample_name=V2) %>% 
  rename(subregion=V5) %>%
  left_join(geneticsex, by="id") %>% 
  left_join(df2, by="sample_name")

df_name <- paste0("clusters_", paste0(gsub("\\_.*","",par)))
assign(df_name, clusters)

mds_plot <- 
  clusters %>% 
  filter(genotype>0.34) %>%
  distinct(id, .keep_all = T) %>% 
  ggplot(aes(y=Dim.2, x=Dim.1, col=as.factor(k)))+
  geom_point()+
  theme_minimal() +
  scale_color_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))+
  theme(
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(), 
    axis.text.x = element_text(hjust=1, size=text_size),
    axis.text.y = element_text(size = text_size),
    axis.title = element_text(size = text_size),
    legend.position = "none",
    legend.text = element_text(size=11),
    legend.title = element_text())+
  labs(x="\nMDS1", y="MDS2\n", 
       title=paste0(gsub("\\_.*","",par), "\n"))

#Add name for specific chromosome 
plot_name <- paste0("mds_plot_", paste0(gsub("\\_.*","",par)))
assign(plot_name, mds_plot)

bodysize_plot <- 
  clusters %>% 
  drop_na(body_PC1) %>% 
  filter(genotype>0.34) %>%
  distinct(id, .keep_all = T) %>% 
  ggplot(aes(y=body_PC1, x=as.factor(k), col=body_PC1))+
  geom_violin(trim = FALSE) +
  stat_summary(
    aes(group = geno),
    geom = "pointrange",
    fun.data = mean_cl_normal,
    fun.args=list(conf.int=0.89),
    size = 1)+
  geom_jitter(width = 0.05,
              height = 0.05,
              alpha = 0.8,
              size=2) +
  theme_minimal() +
  scale_color_viridis_c()+
  #scale_color_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))+
  theme(
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(), 
    axis.text.x = element_text(hjust=1, size=text_size),
    axis.text.y = element_text(size = text_size),
    axis.title = element_text(size = text_size),
    #legend.position = "none",
    legend.text = element_text(size=11),
    legend.title = element_text())+
  labs(x="\nCluster", y="Body size (PC1)\n", 
       title=paste0(gsub("\\_.*","",par), "\n"))

plot_name <- paste0("bodysize_plot_", paste0(gsub("\\_.*","",par)))
assign(plot_name, bodysize_plot)

het_n <- 
  clusters %>% 
  filter(genotype>0.34) %>%
  group_by(k, geno) %>%
  summarise(n=n())

df_name <- paste0("n_het_", paste0(gsub("\\_.*","",par)))
assign(df_name, het_n)

df <-
  clusters %>% 
  group_by(pop, geno, k, region) %>% 
  summarise(n=n())

summary_alleles_cluster <- df %>%
  group_by(pop, geno, k) %>%
  summarise(total_n = sum(n)) %>%
  pivot_wider(names_from = k, values_from = total_n) %>%
  summarise(percentage_cluster1 = sum(`1`, na.rm = TRUE) / sum(`1`, `2`, `3`, na.rm = TRUE),
            percentage_cluster2 = sum(`2`, na.rm = TRUE) / sum(`1`, `2`, `3`, na.rm = TRUE),
            percentage_cluster3 = sum(`3`, na.rm = TRUE) / sum(`1`, `2`, `3`, na.rm = TRUE))


df_map <- summary_alleles_cluster %>%
  left_join(pop_info, by = "pop") %>%
  select(-sample_name) %>% 
  distinct(pop, .keep_all=TRUE)

df_map[is.na(df_map)] <- 0

oceania <-
  rnaturalearth::ne_countries(scale = "large",
               returnclass = "sf",
               continent = "oceania")

anzo <- ggplot(data = oceania) +
    geom_sf(fill = "darkgrey", color = NA) +
    coord_sf(xlim = c(140, 187),
             ylim = c(-50,-12),
             expand = FALSE) +
    theme(
      text=element_text(),
      panel.background = element_rect(fill = background_colour),
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks = element_blank(),
       legend.position = "none",
      plot.margin = margin(0.2, 0.2, 0.2, 0.2, "cm"))+
    geom_scatterpie(aes(x=longitude+0.5, y=latitude+0.5, group = pop, r = 1), 
                    data = df_map, cols=c("percentage_cluster1", 
                                          "percentage_cluster2", 
                                          "percentage_cluster3"), color=NA) + 
    scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))

van_sm <- ggplot(data = oceania) +
    geom_sf(fill = "darkgrey", color = NA) +
    coord_sf(xlim = c(160, 172),
             ylim = c(-25,-12),
             expand = FALSE) +
    theme(
      text=element_text(),
      panel.background = element_rect(fill = background_colour),
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.ticks = element_blank(),
      legend.position = "none",
      plot.margin = margin(0.2, 0.2, 0.2, 0.2, "cm"))+
    geom_scatterpie(aes(x=longitude+0.5, y=latitude+0.5, group = pop, r = 0.3),
                    data = df_map, cols=c("percentage_cluster1",
                                          "percentage_cluster2",
                                          "percentage_cluster3"), color=NA) +
    scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))

chartpie <- 
  anzo+van_sm+plot_layout(guides="collect")

chartpie

plot_name <- paste0("chartpie_plot_", paste0(gsub("\\_.*","",par)))
assign(plot_name, chartpie)
}
```

Only run if chr4A. Decompose matrix and plot neo sex chromosome.
```{r}
beagle_file <- file.path(
  reports_path,
  "localPCA",
  "unpruned",
  "outliers",
  "beagle_by_window",
  "max2_merge10_filter5",
  "chr4A_1257754-chr4A_1261375.beagle"
)

if (!file.exists(beagle_file)) {
  gunzip(paste0(beagle_file, ".gz"), remove = FALSE)
}

outlier <-
  read.table(
    file.path(
      reports_path,
      "localPCA",
      "unpruned",
      "outliers",
      "beagle_by_window",
      "max2_merge10_filter5",
      "chr4A_1257754-chr4A_1261375.beagle"
    ),
    header = T
  ) %>%
  select(-c("allele1", "allele2")) %>%
  as_tibble() %>%
  pivot_longer(starts_with("Ind")) %>%
  rename(sample_name = name) %>%
  rename(genotype = value) %>%
  mutate(sample_name = str_remove(sample_name, "\\.[^.]*$"))

geno <-
  rep(c("AA", "AB", "BB"),
      times = length(outlier$sample_name) / 3)

df <-
  cbind(outlier, geno)

df <- df %>%
  filter(genotype > 0.34) %>%
  group_by(sample_name, marker) %>%
  top_n(1, genotype)

df2 <-
  df %>%
  left_join(pop_info, by = "sample_name") %>%
  left_join(pheno, by = "sample_name") %>%
  separate(marker, into = c("chr", "position")) %>%
  mutate(position = as.numeric(position)) %>%
  unite("marker", chr:position, remove = FALSE)

label <-
  read.table(file.path(data_path, "localPCA/pop_label"))

cov_mat <-
  as.matrix(read.table(
    file.path(
      reports_path,
      "localPCA",
      "unpruned",
      "outliers",
      "cov_by_window",
      "max2_merge10_filter5",
      paste0(
        gsub("\\..*", "", "chr4A_1257754-chr4A_1261375.beagle"),
        ".cov"
      )
    )
  ))

#Do MDS on cov matrix
mds.cor <- (1 - cov_mat) %>%
  cmdscale(k = 2, eig = TRUE)

colnames(mds.cor$points) <- c("Dim.1", "Dim.2")
rownames(mds.cor$points) <-
  label$V3

kmeans_res <-
  kmeans(as.matrix(mds.cor$points[, 1]),
         c(
           min(mds.cor$points[, 1]),
           max(mds.cor$points[, 1])
         ))
k_ss <-
  round(kmeans_res$betweenss / kmeans_res$totss, 3)

k <- as.data.frame(kmeans_res$cluster)
colnames(k) <- "k"

clusters_neosex <-
  label %>%
  select(-V3,-V4) %>%
  cbind(mds.cor$points) %>%
  cbind(k) %>%
  rename(id = V1) %>%
  rename(sample_name = V2) %>%
  rename(subregion = V5) %>%
  left_join(geneticsex, by = "id") %>%
  left_join(df2, by = "sample_name")

mds_neosex <-
  clusters_neosex %>%
  filter(GeneticSex != "NA") %>%
  distinct(id, .keep_all = T) %>%
  ggplot(aes(
    x = Dim.1,
    y = Dim.2,
    col = as.factor(GeneticSex)
  )) +
  geom_point() +
  theme_minimal() +
  scale_color_manual(values = c("#264653", "#e76f51")) +
  theme(
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(hjust = 1, size = text_size),
    axis.text.y = element_text(size = text_size),
    axis.title = element_text(size = text_size),
    legend.position = "right",
    legend.text = element_text(size = 11),
    legend.title = element_text()
  ) +
  labs(x = "\nMDS1", y = "MDS2\n") +
  guides(color = guide_legend(title = "Cluster"))
```


```{r}
fig3 <-
  plot_mds1/(outlier_mds_plots[[1]]+mds_neosex)


fig4 <- 
  (outlier_mds_plots[[2]]+mds_plot_chr27)/(outlier_mds_plots[[3]]+mds_plot_chr28)/(outlier_mds_plots[[4]]+mds_plot_chr29)


fig5<-chartpie_plot_chr27/chartpie_plot_chr28/chartpie_plot_chr29

ggsave(
  fig3,
  filename = file.path(reports_path, 
                       "plots",
                       paste0("fig3.pdf")),
  device = "pdf",width = 10, height=6
)

ggsave(
  fig3,
  filename = file.path(reports_path, 
                       "plots",
                       paste0("fig3.png")),
  device = "png",width = 10, height=6
)

ggsave(
  fig4,
  filename = file.path(reports_path, 
                       "plots",
                       paste0("fig4.pdf")),
  device = "pdf",width = 6, height=8
)

ggsave(
  fig5,
  filename = file.path(reports_path, 
                       "plots",
                       paste0("fig5.pdf")),
  device = "pdf",width = 6, height=8
)

ggsave(
plot_mds_complete[["chr4A"]],
filename = file.path(reports_path, 
                       "plots",
                       paste0("mds_chr4A.pdf")),
  device = "pdf",width = 4.5, height=3
)

ggsave(
mds_neosex,
filename = file.path(reports_path, 
                       "plots",
                       paste0("clusters_chr4A.pdf")),
  device = "pdf",width = 3.5, height=2.5
)

```

LD within and outside inversions
```{r}
list_chr <- c("chr4A", "chr27", "chr28", "chr29", "chr30")
ld_info <- data.frame(
  chr = character(),
  high_ld = numeric(),
  low_ld = numeric(),
  stringsAsFactors = FALSE
)

for (chr in list_chr) {
  # Filtering and arranging in descending order
  if (chr == "chr4A") {
    filtered_data_desc <- mds_scores %>%
      filter(LG == chr) %>%
      arrange(PC1) %>%
      select(pos, PC1)
    closest_col <- "PC1"
  } else {
    filtered_data_desc <- mds_scores %>%
      filter(LG == chr) %>%
      arrange(X2) %>%
      select(pos, X2)
    closest_col <- "X2"
  }

  closest_to_zero_desc <- filtered_data_desc$pos[which.min(abs(filtered_data_desc[[closest_col]] - 0))]

  cat("Positions to calculate LD for", chr, ":\n")
  cat("Low LD predicted:", as.character(closest_to_zero_desc), "\n")

  # Filtering and arranging in ascending order
  if (chr == "chr4A") {
    filtered_data_asc <- mds_scores %>%
      filter(LG == chr) %>%
      arrange(desc(PC1)) %>%
      select(pos, PC1)
  } else {
    filtered_data_asc <- mds_scores %>%
      filter(LG == chr) %>%
      arrange(desc(X2)) %>%
      select(pos, X2)
  }

  first_pos_asc <- pull(filtered_data_asc, pos)[1]

  cat("High LD predicted:", as.character(first_pos_asc), "\n")
  cat("\n")

  # Store the information in the ld_info dataframe
  ld_info <- rbind(ld_info, data.frame(chr = chr, high_ld = as.character(first_pos_asc), low_ld = as.character(closest_to_zero_desc), stringsAsFactors = FALSE))
}

# Print the ld_info dataframe
print(ld_info)
```

Calculate how many heterozygotes in each cluster group
```{r}
df_het <-
  ld_info %>% 
  separate(high_ld, c("high_ld_start", "high_ld_end"), "-") %>% 
  separate(low_ld, c("low_ld_start", "low_ld_end"), "-") %>% 
  filter(chr=="chr27" | chr=="chr28" | chr=="chr29")

LRTs <- 
  fread(file.path(reports_path,
                "GWAS",
                "lrt",
                paste0("body_pc1_gwas_wholegenome_eigen2.lrt0.gz")),
                header = T,
                sep = "\t",
                stringsAsFactors = F) %>%
        mutate(Chromosome = str_replace(Chromosome, "chr", ""))
      
order_chr <- as.character(seq(from = 5, to = 30))
reference = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z")
ref = factor(as.factor(reference), levels = c(c("1", "1A", "2", "3", "4", "4A"), order_chr, "Z"))
LRTs <- LRTs[order(factor(as.factor(LRTs$Chromosome), levels = ref)),]
LRTs$n <- 1:nrow(LRTs)

process_data <- function(data, chromosome, start_pos, end_pos) {
  result <- data %>% 
    mutate(chr=paste0("chr",Chromosome)) %>% 
    filter(Position > start_pos) %>% 
    filter(Position < end_pos) %>% 
    dplyr::select(chr, `high_WT/HE/HO`) %>% 
    separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>% 
    filter(high_WT > 0) %>% 
    filter(HE > 0) %>% 
    filter(HO > 0) %>% 
    mutate(He = as.numeric(HE) / 305)
  
  return(result)
}

results <- list()
for (chromosome in df_het$chr){
  high_range_1 <- 
    df_het %>% filter(chr==chromosome) %>% 
    pull(high_ld_start)
    
  high_range_2 <- 
    df_het %>% filter(chr==chromosome) %>% 
    pull(high_ld_end)
  
  results[[chromosome]] <- process_data(LRTs, chromosome, high_range_1, high_range_2)
}

means <- sapply(results, function(result) mean(result$He))

chr27_high_het <- LRTs %>% 
  mutate(chr = paste0("chr", Chromosome)) %>% 
  filter(chr == "chr27") %>% 
  filter(Position > 1554607) %>% 
  filter(Position < 1557349) %>% 
  dplyr::select(chr, `high_WT/HE/HO`) %>% 
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>% 
  filter(as.numeric(`high_WT`) > 0) %>% 
  filter(as.numeric(HE) > 0) %>% 
  filter(as.numeric(HO) > 0) %>% 
  mutate_all(function(x) as.numeric(as.character(x))) %>% 
  mutate(
    He = as.numeric(HE) / rowSums(across(c("high_WT", "HE", "HO")))
  )

median(chr27_high_het$He)

chr27_low_het <-
  LRTs %>% 
  mutate(chr=paste0("chr",Chromosome)) %>% 
  filter(chr=="chr27") %>%
  filter(Position>597479) %>% 
  filter(Position<602389)  %>% 
  dplyr::select(chr, `high_WT/HE/HO`) %>% 
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>% 
  filter(as.numeric(`high_WT`) > 0) %>% 
  filter(as.numeric(HE) > 0) %>% 
  filter(as.numeric(HO) > 0) %>% 
  mutate_all(function(x) as.numeric(as.character(x))) %>% 
  mutate(
    He = as.numeric(HE) / rowSums(across(c("high_WT", "HE", "HO")))
  )

median(chr27_low_het$He)

chr28_high_het <- LRTs %>%
  mutate(chr = paste0("chr", Chromosome)) %>%
  filter(chr == "chr28") %>%
  filter(Position > 4262245) %>%
  filter(Position < 4266512) %>%
  dplyr::select(chr, `high_WT/HE/HO`) %>%
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>%
  filter(as.numeric(`high_WT`) > 0) %>%
  filter(as.numeric(HE) > 0) %>%
  filter(as.numeric(HO) > 0) %>%
  mutate_all(function(x)
    as.numeric(as.character(x))) %>%
  mutate(He = as.numeric(HE) / rowSums(across(c(
    "high_WT", "HE", "HO"
  ))))

mean(chr28_high_het$He)

chr28_low_het <-
  LRTs %>% 
  mutate(chr=paste0("chr",Chromosome)) %>% 
  filter(chr=="chr28") %>%
  filter(Position>2305134) %>% 
  filter(Position<2306717)  %>% 
  dplyr::select(chr, `high_WT/HE/HO`) %>% 
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>% 
  filter(as.numeric(`high_WT`) > 0) %>% 
  filter(as.numeric(HE) > 0) %>% 
  filter(as.numeric(HO) > 0) %>% 
  mutate_all(function(x) as.numeric(as.character(x))) %>% 
  mutate(
    He = as.numeric(HE) / rowSums(across(c("high_WT", "HE", "HO")))
  ) 
mean(chr28_low_het$He)

chr29_high_het <- LRTs %>%
  mutate(chr = paste0("chr", Chromosome)) %>%
  filter(chr == "chr29") %>%
  filter(Position > 1489919) %>%
  filter(Position < 1495324) %>%
  dplyr::select(chr, `high_WT/HE/HO`) %>%
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>%
  filter(as.numeric(`high_WT`) > 0) %>%
  filter(as.numeric(HE) > 0) %>%
  filter(as.numeric(HO) > 0) %>%
  mutate_all(function(x)
    as.numeric(as.character(x))) %>%
  mutate(He = as.numeric(HE) / rowSums(across(c(
    "high_WT", "HE", "HO"
  ))))

median(chr29_high_het$He)

chr29_low_het <-
  LRTs %>% 
  mutate(chr=paste0("chr",Chromosome)) %>% 
  filter(chr=="chr29") %>%
  filter(Position>60372) %>% 
  filter(Position<63739)  %>% 
  dplyr::select(chr, `high_WT/HE/HO`) %>% 
  separate(`high_WT/HE/HO`, c("high_WT", "HE", "HO"), "/") %>% 
  filter(as.numeric(`high_WT`) > 0) %>% 
  filter(as.numeric(HE) > 0) %>% 
  filter(as.numeric(HO) > 0) %>% 
  mutate_all(function(x) as.numeric(as.character(x))) %>% 
  mutate(
    He = as.numeric(HE) / rowSums(across(c("high_WT", "HE", "HO")))
  ) 
hist(chr28_low_het$He)

median(chr27_high_het$He)
median(chr27_low_het$He)
median(chr28_high_het$He)
median(chr28_low_het$He)
median(chr29_high_het$He)
median(chr29_low_het$He)

```

Another way of measuring Het
```{r}
clusters %>% 
  filter(geno == "AB") %>%
  group_by(geno, k) %>% 
  summarise(n = n())
#chr29: 48% AB in k=2 43153/(43153+30723+14649)
#chr27: 64% in k=3 128004/(128004+11825+59535)

clusters %>% 
  filter(geno == "AA") %>%
  group_by(geno, k) %>% 
  summarise(n = n())
#chr29: 38% AA in k=1 166417/(166417+186085+88377)
#chr27: 63% in k=3 696560/(321561+74270+696560)

clusters %>% 
  filter(geno == "BB") %>%
  group_by(geno, k) %>% 
  summarise(n = n())
#chr29: 45% BB in k=3 31770/(26644+31770+11116)
#chr27: 71% in k=3 153238/(153238+10782+50111)

clusters %>%
  group_by(geno, k) %>%
  summarise(n = n()) %>% 
  ggplot(aes(
    x = as.factor(k),
    y = n,
    fill = as.factor(geno),
    col = as.factor(geno)
  )) + geom_bar(stat = "identity")

df <- 
  clusters %>%
  filter(genotype>0.9) %>% 
  group_by(k, geno) %>%
  summarise(n=n())

summary_alleles_cluster <- 
  df %>%
  group_by(k, geno) %>%
  summarise(total_n = sum(n)) %>%
  pivot_wider(names_from = geno, values_from = total_n) %>%
  summarise(percentage_AA = sum(`AA`, na.rm = TRUE) / sum(`AA`, `AB`, `BB`, na.rm = TRUE),
            percentage_AB = sum(`AB`, na.rm = TRUE) / sum(`AA`, `AB`, `BB`, na.rm = TRUE),
            percentage_BB = sum(`BB`, na.rm = TRUE) / sum(`AA`, `AB`, `BB`, na.rm = TRUE))

#chr27:
#     k    percentage_AA percentage_AB percentage_BB
#     1         0.620         0.264        0.116 
#     2         0.519         0.431        0.0498
#     3         0.652         0.284        0.0641

#chr29
     #  k percentage_AA percentage_AB percentage_BB
     # 1         0.827         0.136        0.0374
     # 2         0.603         0.352        0.0452
     # 3         0.446         0.449        0.105

     #  k percentage_AA percentage_AB percentage_BB
     # 1         0.424         0.426        0.150 
     # 2         0.711         0.220        0.0693
     # 3         0.829         0.131        0.0393
```


