# Prune for LD before GWAS

## Prep data

#### Replace spaces with tabs for ngsLD to read to file without any problem

*Script: 0.0_spaces2tsv.sh* 

*Path:~/sjoh4959/projects/0.0_island_rule/analysis/LD*

*Location: Nesoi* 

```bash
nohup cat /data/Users/Andrea/silvereye/wgs_beagle/raw/wholegenome.beagle | tr ' ' '\t' > /data/Users/Andrea/silvereye/wgs_beagle/raw/whole_genome_ngsLD.beagle &
```

#### Create subsets by chromosome

*Script: 0_create_subsets_4_ngdLD.sh* 

*Path:~/sjoh4959/projects/0.0_island_rule/analysis/LD*

*Location: Nesoi* 



```bash
PATH_INPUT="/data/Users/Andrea/silvereye/wgs_beagle/LD/input/"
PATH_GENOME="/data/Users/Andrea/silvereye/wgs_beagle/raw/"
PATH_HEADER="/data/Users/Andrea/silvereye/wgs_beagle/raw/"
declare -a chromosome_list=()
#Populate the array chromosome_list with strings pasted to numbers
for i in $(seq 30)
do
	tmp=$(echo "chr${i}")
	chromosome_list+=( $tmp )
done

#Add to array those chromosomes with a name that does not follow the pattern above (i.e. chrZ, chr1A...)
chromosome_list+=("chrZ" "chr1A" "chr4A")

#Subset the main file by chromosome and zip
for chromosome in "${chromosome_list[@]}"
do
        nohup cat "${PATH_GENOME}wholegenome.beagle" | grep "${chromosome}_" > "${PATH_INPUT}${chromosome}" &
done
```

#### Add a header to the chromosome files

*Script: 1_create_subsets_4_ngdLD.sh* 

*Path:~/sjoh4959/projects/0.0_island_rule/analysis/LD*

*Location: Nesoi* 

```bash
PATH_INPUT="/data/Users/Andrea/silvereye/wgs_beagle/LD/input/"
PATH_GENOME="/data/Users/Andrea/silvereye/wgs_beagle/raw/"
PATH_HEADER="/data/Users/Andrea/silvereye/wgs_beagle/raw/"
declare -a chromosome_list=()
#Populate the array chromosome_list with strings pasted to numbers
for i in $(seq 30)
do
        tmp=$(echo "chr${i}")
        chromosome_list+=( $tmp )
done

#Add to array those chromosomes with a name that does not follow the pattern above (i.e. chrZ, chr1A...)
chromosome_list+=("chrZ" "chr1A" "chr4A")

#Add header to file
for chromosome in "${chromosome_list[@]}"
do
        nohup cat "${PATH_GENOME}header" "${PATH_INPUT}${chromosome}" | gzip > "${PATH_INPUT}${chromosome}.gz" &
done
```

## Split chromosomes 

```bash
#!/bin/bash

#SBATCH --nodes=1
#SBATCH --job-name=split_chr
#SBATCH --partition=long

cd /data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/input
mkdir subsets

for i in $(seq 5)
do
  	tmp=$(echo "chr${i}")
        chromosome_list+=( $tmp )
done

chromosome_list+=("chr1A" "chr4A")

for chromosome in "${chromosome_list[@]}"
do
	split -l 150000 --numeric-suffixes "${chromosome}_pos" "${chromosome}_pos" 
done

for chromosome in "${chromosome_list[@]}"
do
	zcat "${chromosome}.gz" | tail -n +2 > tmp
    split -l 150000 --numeric-suffixes tmp "${chromosome}_"
    rm tmp
    mv *"${chromosome}_0"* subsets
    mv *"${chromosome}_1"* subsets
    #cd subsets
    #gzip "${chromosome}"
    #cd ../
done

cd /data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/input/subsets

for i in *_1*
do	
	cat header $i > tmp
	gzip tmp 
	mv tmp.gz "${i}.gz"
	mv "${i}.gz" ..
done

gzip < subsets/header > chr1_10_1.gz
cat chr1_10.gz >> chr1_10_1.gz
rm chr1_10.gz
mv chr1_10_1.gz chr1_10.gz
```

## Calculate LD by chromosome

We used ngsLD (REF) to prune those sites that were in LD > X. 

*Script: 0_run_ngsLD.sh* 

*Path: /data/zool-zost/sjoh4959/projects/0.0_island_rule/slurm/LD*

*Location: ARC* 

```bash
#!/bin/bash

#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --job-name=ngsLD
#SBATCH --partition=long

module purge
module load ngsLD/2020

PATH_INPUT="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/input/"
PATH_OUTPUT="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/output/"
N_IND=377
N_THREADS=40
MAX_KB_DIST=0
MAX_SNP_DIST=0

declare -a chromosome_list=()
#Populate the array chromosome_list with strings pasted to numbers
for i in $(seq 30)
do
  	tmp=$(echo "chr${i}")
        chromosome_list+=( $tmp )
done

#Add to array those chromosomes with a name that does not follow the pattern above (i.e. chrZ, chr1A...)
chromosome_list+=("chrZ" "chr1A" "chr4A")

#Create POS file
for chromosome in "${chromosome_list[@]}"
do
        zcat "${PATH_INPUT}${chromosome}.gz" | cut -f -1 | awk '{if(NR>1)print}' > pos_tmp1 
        cat pos_tmp1 | sed -r 's/_/ /g' > pos_tmp2
        cat pos_tmp2 | tr ' ' '\t' > "${PATH_INPUT}${chromosome}_pos"
        rm pos_tmp1
        rm pos_tmp2
done

#Run ngsLD
for chromosome in "${chromosome_list[@]}"
do
  	N_SITES=$(cat "${PATH_INPUT}${chromosome}_pos" | wc -l)
  	ngsLD --geno "${PATH_INPUT}${chromosome}.gz" --probs --n_ind  $N_IND  --n_sites $N_SITES --out "${PATH_OUTPUT}${chromosome}.ld" --n_threads $N_THREADS --max_kb_dist 0 --max_snp_dist 0 --extend_out --pos "${PATH_INPUT}${chromosome}_pos" --rnd_sample 0.1
done
```

## Test for LD decay

Activate conda environment. We’ve also created an R project with an associated environment that contains the dependencies necessary to use fit_LDdecay.R

```bash
conda activate LD
```

Create a list for each chromosome or chunk of chromosomes

```bash
for i in *ld 
	do touch $i.list | echo $i > $i.list 
done
```

```bash
PATH_INPUT="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/LD/"
PATH_OUTPUT="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/LD/plots/"
PATH_SCRIPT="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/LD/"

for i in *list    
do 
        nohup Rscript --vanilla --slave "${PATH_SCRIPT}fit_LDdecay.R" --ld_files "${PATH_INPUT}$i" --out "${PATH_OUTPUT}$i.pdf" --max_kb_dist 100 &
done
```





## Obtain sites for pruning

*Script: 1_pruneLD_chr.sh* 

*Path: /data/zool-zost/sjoh4959/projects/0.0_island_rule/slurm/LD*

*Location: ARC* 

```bash
#!/bin/bash

#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --job-name=ngsLD
#SBATCH --partition=long

module purge
module load ngsLD/2020

PATH_PRUNE="/data/zool-zost/sjoh4959/projects/0.0_island_rule/src/others/ngsLD/scripts/prune_graph.pl"
PATH_INPUT="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/output/"
PATH_OUT="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/output/pruned_sites/"
MIN_WEIGHT=0.5
MAX_KB_DIST=5

declare -a chromosome_list=()
#Populate the array chromosome_list with strings pasted to numbers
for i in $(seq 30)
do
  	tmp=$(echo "chr${i}")
        chromosome_list+=( $tmp )
done

#Add to array those chromosomes with a name that does not follow the pattern above (i.e. chrZ, chr1A...)
chromosome_list+=("chrZ" "chr1A" "chr4A")

#
for chromosome in "${chromosome_list[@]}"
do
  	nohup perl $PATH_PRUNE --in_file "${PATH_INPUT}${chromosome}.ld" --max_kb_dist $MAX_KB_DIST --min_weight $MIN_WEIGHT --out "${PATH_OUT}${chromosome}_pruned.ld"
done
```

## LD pruning and concatenation into one single whole genome file

```bash
#!/bin/bash

#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --job-name=LD pruning
#SBATCH --partition=medium

PATH_LD_RESULTS="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/output/pruned_sites/"
PATH_INPUT_FILE="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/input/"
PATH_OUTPUT="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/LD/output/pruned_sites/"
PATH_DATA="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/"
PATH_DATA="/data/zool-zost/sjoh4959/projects/0.0_island_rule/data/"

declare -a chromosome_list=()
#Populate the array chromosome_list with strings pasted to numbers
for i in $(seq 30)
do
  	tmp=$(echo "chr${i}")
        chromosome_list+=( $tmp )
done

#Add to array those chromosomes with a name that does not follow the pattern above (i.e. chrZ, chr1A...)
chromosome_list+=("chrZ" "chr1A" "chr4A")

#Select independent sites
for chromosome in "${chromosome_list[@]}"
do
	cat "${PATH_LD_RESULTS}${chromosome}_pruned.ld" | tr ':' '\t' | cut -f2 > sites2keep
	#re="^($(paste -sd '|' sites2keep))$"
	zcat "${PATH_INPUT_FILE}${chromosome}.gz" | grep "^${chromosome}_" | tr '_' '\t' > tmp
	#awk -F '\t' -v re="${re//./\\.}" 'NR==1 || $2 ~ re' tmp > tmp2
	grep -w -F -f sites2keep tmp > tmp2
	#Concatenate first and second columns with "_"
	cat tmp2 | cut -f1,2 | tr '\t' '_' > first_col
	paste first_col tmp2 | cut -f2,3 --complement > tmp3
	#Add header
	cat "${PATH_DATA}header" tmp3 > "${PATH_OUTPUT}${chromosome}_pruned.beagle"
	#clean
	rm sites2keep first_col tmp tmp2 tmp3
done

#Concatenate pruned chromosomes into pruned a single whole genome file
cd ${PATH_OUTPUT}

cat chr16_pruned.beagle | head -n 1 > wholegenome_pruned.beagle

#Sort list of files in natural order
readarray -d '' entries < <(printf '%s\0' *pruned.beagle | sort -zV)

#Concatenate the files
for entry in "${entries[@]}" 
do 
	cat $entry | tail -n +2 -q >> wholegenome_pruned.beagle
done

#Move whole genome to final destination
mv wholegenome_pruned.beagle $PATH_DATA
```



# Genome-wide Association Study

## Check phenotype normality and subset BEAGLE file by phenotype

*Script: 0_sorting_phenotypes.Rmd* 

*Path: notebooks*

*Location: Project* 

```{r setup}
# Knitr settings: 
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
options(scipen = 999)
```

```{r}
source("./src/0.0_island_rule_source.R")
```

**Read data**

```{r}
pheno <- 
  read_csv(file.path(data_path, "raw/phenotypes+SMC2+AE.csv")) %>% 
  mutate(wing_weight=Wing/Weight) %>% 
  mutate(wing_weight=ifelse(wing_weight=="Inf", NA, wing_weight)) %>% 
  mutate(tarsus_weight=Tarsus/Weight) %>% 
  mutate(tarsus_weight=ifelse(tarsus_weight=="Inf", NA, tarsus_weight)) #%>%
  #mutate(id=firstup(id)) 

pheno[pheno=="<NA>"]=NA
```

**PCA prep for GWAS with a) head, wing, tarsus, tail b) bill measurements**

```{r}
body_pca <-
  princomp(~Wing+
             Tarsus+
             HeadLength,
           data=pheno,
           scores=TRUE,
           cor=TRUE,
           na.action=na.exclude)

colnames(body_pca$scores) <- 
  c("body_PC1", "body_PC2", "body_PC3")

bill_pca <-
  princomp(~Bill_length_posterior+
             Bill_depth_anterior+
             Bill_width_anterior,
           data=pheno,
           scores=TRUE,
           cor=TRUE,
           na.action=na.exclude)

colnames(bill_pca$scores) <- 
  c("bill_PC1", "bill_PC2", "bill_PC3")

pheno <- 
  cbind(pheno,body_pca$scores) %>% 
  cbind(bill_pca$scores)

write.csv(pheno, file.path(data_path, "dfs/pheno_pca.csv"), row.names = F)
```

**Plot trait data per population**

```{r}
plot_phenotypes <- function(dataset, trait, xaxis_name, color){
  plot = dataset %>%
    ggplot(aes(y = .data[[trait]],
               x = reorder(pop, .data[[trait]], mean, na.rm=TRUE),
               color=color)) +
    stat_summary(
      aes(group = pop),
      geom = "pointrange",
      fun.data = mean_cl_boot,
      size = 1
    ) +
    geom_jitter(
      width = 0.1,
      height = 0.1,
      alpha = 0.25
    ) +
    scale_color_manual(values=color)+
    theme_minimal() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major.x = element_blank(),
      axis.text.x = element_text(angle = 45, hjust=1, size=text_size),
      axis.text.y = element_text(size = text_size),
      axis.title = element_text(size = text_size),
      axis.title.x = element_blank(),
      legend.position = "none") +
     labs(
       y = paste0(xaxis_name, "\n"))
    return(plot)
}

trait_list = colnames(pheno)[c(13:16, 23, 28:29)]
output_plot <- list()
for (trait in trait_list){
  output_plot[[trait]] <- plot_phenotypes(pheno, trait, as.character(trait), color="#3d3d3d")
}

trait_plot <- wrap_plots(output_plot)

ggsave(
  "trait_plot.pdf",
  trait_plot,
  path = figures_path,
  device = "pdf",
  width = 30,
  height = 10,
  dpi = 400
)

pc_list = colnames(pheno)[c(30:31, 32:33)]
output_plot_pc <- list()
for (trait in pc_list){
  output_plot_pc[[trait]] <- plot_phenotypes(pheno, trait, as.character(trait), color="#3d3d3d")
}

pc_plot <- wrap_plots(output_plot_pc)

ggsave(
  "pc_plot.pdf",
  pc_plot,
  path = figures_path,
  device = "pdf",
  width = 12,
  height = 6,
  dpi = 400
)

for (i in 1:length(output_plot)){
ggsave(
  paste0("plot_", names(output_plot)[i], ".pdf"),
  output_plot[[i]],
  path = figures_path,
  device = "pdf",
  width = 6,
  height = 3,
  dpi = 400
)
}
```

**Summary stats phenotypes**

```{r}
summary_stats <- list()
total_sample_size <- list()
for (trait in colnames(pheno[c(13:23)])) {
  summary_stats[[trait]] <-
    group_by(pheno, pop) %>%
    drop_na(trait) %>%
    summarise(
      count = n(),
      mean = mean(get(trait), na.rm = TRUE),
      max = max(get(trait), na.rm = TRUE),
      min = min(get(trait), na.rm = TRUE),
      sd = sd(get(trait), na.rm = TRUE),
    )
  total_sample_size[[trait]] <-
    pheno %>%
    dplyr::select(trait) %>% 
    drop_na() %>% 
    count()
}

#Save list in an Excel workbook with each list on a sheet
for (sublist in 1:length(summary_stats)){
  write.xlsx(summary_stats[sublist], 
             file=file.path(figures_path, "summary_stats_pheno.xlsx"), 
             sheetName=paste(sublist), 
             row.names=FALSE,
             append=T)
}
```

**Check normality in phenotypes **

```{r, dev='pdf', cache = TRUE, message=FALSE, include=TRUE}
trait_list <- c(trait_list, "body_PC1", "body_PC2", "bill_PC1", "bill_PC2")
plot_normality <- function(dataset, traits, xaxis_name) {
  plot = dataset %>%
    ggplot(aes(x = .data[[trait]])) +
    geom_density() +
    theme_minimal() +
    labs(x = paste0(xaxis_name, "\n"))
  return(plot)
}

plot_qqnormality <- function(dataset, traits, yaxis_name) {
  plot = dataset %>%
    ggplot(aes(sample = .data[[trait]])) +
    stat_qq() + stat_qq_line()+
    theme_classic() +
    labs(y = paste0(yaxis_name, "\n"))
  return(plot)
}

trait_list <- c(trait_list, "body_PC2")
output_norm_plot <- list()
for (trait in trait_list) {
  output_norm_plot[[trait]] <- plot_normality(pheno, trait, as.character(trait))
}

output_qqnorm_plot <- list()
for (trait in trait_list) {
  output_qqnorm_plot[[trait]] <- plot_qqnormality(pheno, trait, as.character(trait))
}

norm_plots <- wrap_plots(output_norm_plot)
qqnorm_plots <- wrap_plots(output_qqnorm_plot)

ggsave(
  "norm_plots.pdf",
  norm_plots,
  path = figures_path,
  device = "pdf",
  width = 8,
  height = 6,
  dpi = 400
)

ggsave(
  "qqnorm_plots.pdf",
  qqnorm_plots,
  path = figures_path,
  device = "pdf",
  width = 8,
  height = 6,
  dpi = 400
)
```

**Generate subsets for sequence removal in BEAGLE file**

```{r}
samples2keep <- list()
for (trait in trait_list){
  samples2keep[[trait]] <- 
    pheno %>% 
      dplyr::select(sample_name, id, pop, trait) %>% 
      drop_na()
}
##SAMPLES TO KEEP##
#Save list in an Excel workbook with each list on a sheet. 
for (sublist in 1:length(samples2keep)){
  write.xlsx(samples2keep[sublist], 
             file=file.path(figures_path,"samples2keep.xlsx"), 
             sheetName=paste(sublist), 
             row.names=FALSE,
             append=T)
}

##SAMPLES TO REMOVE##
#Generate sequence from Ind0 to Ind187  
numbers <- seq(0,387)
m <- c()
for (number in numbers){
  x <- paste0("Ind",number,collapse="")
  m <- c(m,x)
}
#Check those samples that are in the main subset but not in the just generated sequence
samples2remove <- list()
for (sublist in samples2keep){
  samples2remove[[colnames(sublist[4])]] <- 
    as.data.frame(m) %>% 
    filter(m %!in% sublist$id)
}

for (sublist in 1:length(samples2remove)){
  write.xlsx(samples2remove[sublist], 
             file=file.path(figures_path,"samples2remove.xlsx"), 
             sheetName=paste(sublist), 
             row.names=FALSE,
             append=T)
}


for (trait in trait_list){
  tmp <- 
    pheno %>% 
      dplyr::select(id, trait) %>%
      drop_na() %>% 
      dplyr::select(id, trait)
    write_tsv(tmp, file.path(subset_ind_path, paste0(as.character(trait),".tsv")), col_names=F)
}

for (trait in trait_list){
  tmp <- 
    pheno %>% 
      dplyr::select(trait) %>% 
      drop_na()
    write_tsv(tmp, file.path(subset_pheno_path, paste0(as.character(trait),".tsv")), col_names=F)
}
```

## Subset BEAGLE files

*Script: 0_sorting_phenotypes.Rmd* 

*Path: notebooks*

*Location: Project* 

```bash
PATH_OUT="/data/Users/Andrea/silvereye/wgs_beagle/gwas_input/"
PATH_SCRIPT="~/sjoh4959/projects/0.0_island_rule/src/myscripts/"
PATH_LIST="/data/Users/Andrea/silvereye/wgs_beagle/list_samples2keep/"
PATH_GENOME="/data/Users/Andrea/silvereye/wgs_beagle/raw/"

declare -a trait_list=("wing" "tail" "tarsus" "head" "weight" "body_pc1" "body_pc2" "bill_pc1" "bill_pc2") 

for trait in "${trait_list[@]}"
do
        python ""${PATH_SCRIPT}subset_beagle.py" --beagle "${PATH_GENOME}wholegenome_pruned.beagle" --samples "${PATH_LIST}${trait}.tsv" --out "${PATH_OUT}${trait}.beagle"
done

python subset_beagle.py --beagle "${PATH_GENOME}wholegenome_pruned.beagle" --samples "${PATH_LIST}wing.tsv" --out "${PATH_OUT}wing.beagle"
```

## Calculate population structure with PCAngsd

*Script: subset_beagle.py*

*Path: 0.0_island_rule/src/myscripts*

*Location: Nesoi*

```bash
PATH_PCANGSD="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/src/others/pcangsd/"
PATH_BEAGLE="/data/Users/Andrea/silvereye/wgs_beagle/gwas_input/"
PATH_OUT="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/analysis/PCAngsd/out/"
N_THREADS=56
declare -a trait_list=("wing" "tail" "tarsus" "head" "weight" "body_pc1" "body_pc2", "bill_pc1", "bill_pc2")

for trait in "${trait_list[@]}"
do
       nohup python "${PATH_PCANGSD}pcangsd.py" -beagle "${PATH_BEAGLE}${trait}.beagle.gz" -admix -dosage_save -out "${PATH_OUT}${trait}_pca_out" -threads $N_THREADS &
done
```

It’s necessary to calculate eigenvalues so the pop structure is well taken into account. Here I’ve picked the first 10 components but with 4-5 it’s probably enough. 

write.table(eigen(as.matrix(read.table("bill_pc1_chr1A_pca_out.cov")))$vectors[,1:10], "bill_pc1_chr1A_pca_1", row.names=F)



## GWAS with ANGSD

<u>**#NEED TO GENERATE A WHOLEGENOME_PRUNED.FAI. Now we only have wholegenome.fai (!!!)**</u>

```bash
PATH_ANGSD="/home/zoo/sjoh4959/sjoh4959/programs/angsd/angsd"
PATH_GENOME="/data/Users/Andrea/silvereye/wgs_beagle/gwas_input/"
PATH_PHENOTYPES="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/data/phenotypes/"
PATH_FAI="/data/Users/Andrea/silvereye/ref_genome/wholegenome_pruned.fai"
PATH_OUT="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/analysis/GWAS_angsd/output/"
PATH_COV="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/analysis/PCAngsd/out/"
N_THREADS=56

#declare -a trait_list=("wing" "tail" "tarsus" "head" "weight" "body_pc1" "body_pc2" "bill_pc1" "bill_pc2")

for trait in "${trait_list[@]}"
do
	nohup $PATH_ANGSD -doMaf 4 -beagle "${PATH_GENOME}${trait}.beagle" -doAsso 4 -yQuant "${PATH_PHENOTYPES}${trait}.tsv" -out "${PATH_OUT}${trait}_gwas_wholegenome" -fai $PATH_FAI -cov "${PATH_COV}${trait}" -Pvalue 1 &
done
```

## Read output and do a Manhattan plot

```
```



## Select outliers 

```bash
grep -w -f bill_pc1_GWAS_outliers.csv wholegenome_pruned.beagle > bill_pc1_outliers.beagle
```

Then open in R with script process_gwas_outliers.Rmd

# Genome annotation



# Explore inversion in Chr4A

Get the windows with extreme values and save them on a file

```R
list_outliers <- big_data %>% filter(PC1 < -0.3) %>% dplyr::select(file_name)
write.csv(list_outliers, "list_outliers.csv", row.names=F)
```

Copy all those interesting windows into a new folder for upload to the cluster (Nesoi)

```bash
cd reports/localPCA/beagle_by_window/chr4A
mkdir inversion_pca
while read p; do cp $p inversion_pca/; done < list_outliers.csv
cd inversion_pca
scp * sjoh4959@zoo-nesoi.zoo.ox.ac.uk:/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/localPCA/inversion_chr4A
```

Once in Nesoi, create an empty file, unzip and print all lines except for header into the empty file. Then create a header file and concat header and all printed files into a zipped file that we’ll use for PCA

```
touch output
for i in *.beagle.gz; do zcat $i | tail -n+2 >> output;done
zcat window_chr4A_10019120-10042158_100_snps.beagle.gz | head -n 1 > header
cat header output | gzip > inversions_chr4A.beagle.gz
rm output header #to save space :)
```

```bash
PATH_PCANGSD="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/src/others/pcangsd/"
PATH_BEAGLE="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/localPCA/inversion_chr4A/"
PATH_OUT="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/localPCA/inversion_chr4A/"
N_THREADS=56

nohup python "${PATH_PCANGSD}pcangsd.py" -beagle "${PATH_BEAGLE}inversions_chr4A.beagle.gz" -admix -out "${PATH_OUT}inversions_chr4A_pca_out" -threads $N_THREADS &
```

# BEAGLE TO VCF



```bash
#############
###MARKERS###
#############

#Create a file with individuals that will be used as header for the markers file
zcat window_chr30_25104-28571_100_snps.beagle.gz | cut -f 4- | head -n 1 > tmp0

while read line; do
	echo $line | tr ' ' '\n' | uniq | tr '\n' '   ' > ind
	done < tmp0

sed -i '1s/^/I /' ind
sed -i -e '$a\' ind

rm tmp0

#Create a temporary file with the first three lines of the BEAGLE file
zcat window_chr30_25104-28571_100_snps.beagle.gz | awk '{print $1,$2,$3}' > tmp0
#Create three files: one with the marker, the second with the alleleA but changing the numeric code to a ACTG code
cat tmp0 | cut -f1 -d' ' > column0 #marker line
cat tmp0 | awk '{print $2}' | sed 's/0/A/' | sed 's/1/C/' | sed 's/2/T/' | sed 's/3/G/' > column1 #alleleA
cat tmp0 | awk '{print $3}' | sed 's/0/A/' | sed 's/1/C/' | sed 's/2/T/' | sed 's/3/G/' > column2 #alleleB
#Combine all columns into a temporary file
paste column0 column1 column2 > tmp1
#Remove the header of the temporary file
cat tmp1 | tail -n+2 > tmp2
#Add the ind file as header: it contains an I to indicate that these are this line corresponds to individuals and then all individuals
cat ind tmp2 > tmp3
#Subtitute tabs for three spaces
sed -e 's/\t/   /g' tmp3 > markers
#Clean up
rm column0 column1 column2 tmp0 tmp1 tmp2 tmp3

############
###BEAGLE###
############

#Remove the three first columns
zcat window_chr30_25104-28571_100_snps.beagle.gz | cut -f 4-  > tmp4
#Add three columns: marker, 
cat markers | tail -n+2 | awk '{print $1,$2,$3}' > tmp2
touch header
printf 'marker   alleleA   alleleB\n' > header
cat header tmp2 > tmp3
paste tmp3 tmp4 > beagle
rm tmp2 tmp3 tmp4

############
###CHROMO###
############

zcat window_chr30_25104-28571_100_snps.beagle.gz | cut -f 1  > tmp5
sed 's/_.*//' tmp5 | tail -n+2 > chromosome
rm tmp5

#RUN BEAGLE2VCF

java -jar beagle2vcf.jar chromosome beagle markers NA 
```

# OUTLIER TO TABLE

```bash
#transpose ind line to column removing first three columns
cat header | cut -f2- | cut -f2- | cut -f2- | tr '\t' '\n' > tmp
#find position and remove three first columns then concatenate
position="98969775"
cat wholegenome_pruned.beagle | grep $position | cut -f2- | cut -f2- | cut -f2- | tr '\t' '\n' > tmp1
paste tmp tmp1 > outlier_chr3

```

# OUTLIER MDS REGIONS TO COVARIANCE MATRIX

```bash
CHOSEN_FILE="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/localPCA/outlier_mds/cluster_outlier_allMDS_max4_merge20_filter4_sdLim4.pos"


num_pca=$(wc -l "$CHOSEN_FILE" | cut -d " " -f 1)

path_beagle_all="/data/Users/Andrea/silvereye/wgs_beagle/raw/"
path_beagle_subset="/data/Users/Andrea/silvereye/projects_data/0.0_island_rule/analysis/localPCA/outlier_mds/"
PATH_PCANGSD="/home/zoo/sjoh4959/sjoh4959/projects/0.0_island_rule/src/others/pcangsd/"


for i in $(seq $num_pca)
	do
	start=$(cat "$CHOSEN_FILE" | head -"$i" | tail -1 | cut -f 1)
	stop=$(cat "$CHOSEN_FILE" | head -"$i" | tail -1 | cut -f 2)

	echo "splitting beagle $start $stop"
	#Search from start to stop with awk including both lines. Note that start and stop are declared variables so we need to declare the pattern with -v pat="$start" and -v pat2="$stop" and then call it within awk as $0 ~ pat or $0 ~ pat2
	awk -v pat="$start" -v pat2="$stop" '$0 ~ pat{flag=1} $0 ~ pat2{flag=0} flag' ${path_beagle_all}/wholegenome_pruned.beagle > tmp
	
	cat "${path_beagle_all}/header" tmp > "${path_beagle_subset}${start}-${stop}.beagle" 
	rm tmp
	gzip "${path_beagle_subset}${start}-${stop}.beagle"

	echo "pcangsd"
	nohup python "${PATH_PCANGSD}pcangsd.py" -threads 20 -beagle "${path_beagle_subset}${start}-${stop}.beagle.gz" -o "${path_beagle_subset}${start}-${stop}" &
	
done

```

